src/kerberos.cc

```diff
@@ -2,180 +2,191 @@
-    Nan::SetPrototypeMethod(tpl, "wrap", WrapData);
-    Nan::SetPrototypeMethod(tpl, "unwrap", UnwrapData);
-
-    v8::Local<v8::ObjectTemplate> itpl = tpl->InstanceTemplate();
-    itpl->SetInternalFieldCount(1);
-
-    Nan::SetAccessor(itpl, Nan::New("username").ToLocalChecked(), KerberosClient::UserNameGetter);
-    Nan::SetAccessor(itpl, Nan::New("response").ToLocalChecked(), KerberosClient::ResponseGetter);
-    Nan::SetAccessor(
-        itpl, Nan::New("responseConf").ToLocalChecked(), KerberosClient::ResponseConfGetter);
-    Nan::SetAccessor(
-        itpl, Nan::New("contextComplete").ToLocalChecked(), KerberosClient::ContextCompleteGetter);
-
-    constructor.Reset(Nan::GetFunction(tpl).ToLocalChecked());
-    Nan::Set(target,
-             Nan::New("KerberosClient").ToLocalChecked(),
-             Nan::GetFunction(tpl).ToLocalChecked());
-}
-
-v8::Local<v8::Object> KerberosClient::NewInstance(krb_client_state* state) {
-    Nan::EscapableHandleScope scope;
-    v8::Local<v8::Function> ctor = Nan::New<v8::Function>(KerberosClient::constructor);
-    v8::Local<v8::Object> object = Nan::NewInstance(ctor).ToLocalChecked();
-    KerberosClient* class_instance = new KerberosClient(state);
-    class_instance->Wrap(object);
-    return scope.Escape(object);
-}
-
-KerberosClient::KerberosClient(krb_client_state* state) : _state(state) {}
-
-krb_client_state* KerberosClient::state() const {
-    return _state;
-}
+namespace node_kerberos {
+
+using namespace Napi;
+
+namespace {
+struct InstanceData {
+    Reference<Function> KerberosClientCtor;
+    Reference<Function> KerberosServerCtor;
+};
+
+static constexpr napi_property_attributes writable_and_configurable =
+    static_cast<napi_property_attributes>(
+        static_cast<int>(napi_writable) | static_cast<int>(napi_configurable));
 
-NAN_GETTER(KerberosClient::UserNameGetter) {
-    KerberosClient* client = Nan::ObjectWrap::Unwrap<KerberosClient>(info.This());
-    (client->state()->username == NULL)
-        ? info.GetReturnValue().Set(Nan::Null())
-        : info.GetReturnValue().Set(Nan::New(client->state()->username).ToLocalChecked());
+inline String NewMaybeWideString(Env env, const char* str) {
+    return String::New(env, str);
+}
+#ifdef _WIN32
+inline String NewMaybeWideString(Env env, const WCHAR* str) {
+    static_assert(sizeof(std::wstring::value_type) == sizeof(std::u16string::value_type),
+        "wstring and u16string have the same value type on Windows");
+    std::wstring wstr(str);
+    std::u16string u16(wstr.begin(), wstr.end());
+    return String::New(env, u16);
+}
+#endif
 }
 
-NAN_GETTER(KerberosClient::ResponseGetter) {
-    KerberosClient* client = Nan::ObjectWrap::Unwrap<KerberosClient>(info.This());
-    (client->state()->response == NULL)
-        ? info.GetReturnValue().Set(Nan::Null())
-        : info.GetReturnValue().Set(Nan::New(client->state()->response).ToLocalChecked());
+std::string ToStringWithNonStringAsEmpty(Napi::Value value) {
```

Comment:
I know this is a carryover from the original implementation, but I think it might be bording on buggy behavior to just convert non-strings into the empty string. Do you think we should signal something back to the user? It might be a moot point if we do a good enough job validating inputs in the JS part of the module

Reply:
This is definitely something that’s being relied on, in particular the fact that missing options are equivalent to an empty string.

Reply:
It’s probably something that I’d handle in the JS layer at some point, instead of here (so that the bindings always receive options objects that are fully filled out) – but it is a not a bad thing per se that “this option is missing” is translated to “as a string, it is empty”, I think.

Reply:
OK, I'll defer to you, @nbbeeken, and @dariakp on whether you'd like to make a tracking ticket for this

Reply:
Done! NODE-3647